### PR TITLE
feat: added a Card2D widget to UI

### PR DESCRIPTION
**Pull request by antrikshmisri**
This adds a card2d widget as mentioned in issue #397.
Here is a screenshot for reference:-
Screenshot (176)

### CODE REVIEW

**Review by skoudoro — COMMENTED** [review]
Hi @antrikshmisri,

Thank you for doing this! 

Before any review, you will need to add unit tests and a tutorial.
Please look at the other widget for more information.  All lines of the code need to be tested and the tutorial easy to follow. 

Thank you and looking forward to testing + tutorials

**Comment by skoudoro** [comment]
Also, all [pep8 issues](https://github.com/fury-gl/fury/pull/398#issuecomment-807334720) need to be addressed!

**Comment by antrikshmisri** [comment]
> Hi @antrikshmisri,
> 
> Thank you for doing this!
> 
> Before any review, you will need to add unit tests and a tutorial.
> Please look at the other widget for more information. All lines of the code need to be tested and the tutorial easy to follow.
> 
> Thank you and looking forward to testing + tutorials

Hey @skoudoro , will add the tests and tutorial soon.

**Comment by antrikshmisri** [comment]
@skoudoro I have added the tests and tutorial for this widget. Do let me know if I need to add/change anything

File: docs/tutorials/02_ui/viz_card.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+"""
+===============
+Buttons & Text
+===============
+
+This example shows how to use the UI API. We will demonstrate how to create
+panel having multiple cards on them.
+
+First, some imports.
```
Comment (Nibba2018) [inline docs/tutorials/02_ui/viz_card.py:10]: Edit and update the comment according to your tutorial.

File: docs/tutorials/02_ui/viz_card.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+"""
+from numpy.core.fromnumeric import size
+from fury import ui, window
+from fury.data import read_viz_icons, fetch_viz_icons
+
```
Comment (Nibba2018) [inline docs/tutorials/02_ui/viz_card.py:15]: Remove the `numpy.core.fromnumeric...` import as its not being used in your script.

File: docs/tutorials/02_ui/viz_card.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+###############################################################################
+# Now that all the elements have been initialised, we add them to the show
+# manager.
+
+current_size = (1000, 1000)
+show_manager = window.ShowManager(size=current_size,
+                                  title="FURY Card Example")
+
+show_manager.scene.add(panel)
+show_manager.start()
```
Comment (Nibba2018) [inline docs/tutorials/02_ui/viz_card.py:59]: In this part of the tutorial you need to add a `interactive` flag which will make sure that the tutorial should start only when the `interactive` flag is set to `True` and should save the visualisation image when it is set to `False`.

Look at how the other tutorials have implemented it.

File: fury/ui.py [hunk 4]
```diff
@@ -5240,3 +5240,165 @@ def _set_position(self, coords):
         pass
         # self.actor.SetPosition(*coords)
         # self.container.SetPosition(*coords)
+
+
+class Card2D(UI):
+
+    """Card element to show image and related text
+    """
+
```
Comment (Nibba2018) [inline fury/ui.py:5249]: Please add the `Attributes` section here which explains what fundamental UI elements are used to render this particular UI component. Look at the `ComboBox2D` code for e.g.

File: fury/ui.py [hunk 5]
```diff
@@ -5240,3 +5240,165 @@ def _set_position(self, coords):
...
+        self,
+        image_path,
+        body,
+        title,
+        padding=2,
+        position=(0, 0),
+        size=(400, 400),
+        image_scale=0.5,
+        text_color=(0., 0., 0.),
+            ):
```
Comment (Nibba2018) [inline fury/ui.py:5260]: * Its better if the arguments are formatted horizontally instead. (As far as PEP8 allows you to)
* Add default values for `body` and `title`.
* Add additional arguments to:
    *  make the component draggable (or not) 
    * add color/opacity to the bg panel
    * Add color to body and title separately

File: fury/ui.py [hunk 6]
```diff
@@ -5240,3 +5240,165 @@ def _set_position(self, coords):
...
+            UI component
+        size : (int, int)
+            Width and height of the pixels of this UI component.
+        font_size: (int)
+            size of the font
+        text_color: (float , float , float)
+            color of the text
+        color: (float, float, float)
+            Background color of card
+        """
```
Comment (Nibba2018) [inline fury/ui.py:5285]: No need to add `()` for scalar variables.

File: fury/ui.py [hunk 7]
```diff
@@ -5240,3 +5240,165 @@ def _set_position(self, coords):
...
+        self.card_size = size
+        self.padding = padding
+        self.text_color = text_color
+        self.text_scale = np.clip(1 - image_scale, 0, 1)
+        self._image_size = (self.card_size[0], self.card_size[1] *
+                            np.clip(image_scale, 0, 1))
+
+        super(Card2D, self).__init__()
+        self.panel.position = position
+
```
Comment (Nibba2018) [inline fury/ui.py:5299]: `self.position = position` instead of `self.panel.position`

File: fury/ui.py [hunk 8]
```diff
@@ -5240,3 +5240,165 @@ def _set_position(self, coords):
...
+        self.panel.add_element(self.image, _img_coords)
+        self.panel.add_element(self.title_box, _title_coords)
+        self.panel.add_element(self.body_box, _text_coords)
+
+    def _get_actors(self):
+        """ Get the actors composing this UI component.
+        """
+
+        return self.panel.actors + self.image.actors \
+            + self.title_box.actors + self.body_box.actors
```
Comment (Nibba2018) [inline fury/ui.py:5340]: `self.panel.actors` should suffice.

File: fury/ui.py [hunk 9]
```diff
@@ -5240,3 +5240,165 @@ def _set_position(self, coords):
...
+        Parameters
+        ----------
+        scene : scene
+        """
+
+        self.panel.add_to_scene(_scene)
+        self.image.add_to_scene(_scene)
+        self.title_box.add_to_scene(_scene)
+        self.body_box.add_to_scene(_scene)
+
```
Comment (Nibba2018) [inline fury/ui.py:5353]: No need to add these components to the scene as you have already added them to the panel.

File: fury/ui.py [hunk 10]
```diff
@@ -5240,3 +5240,165 @@ def _set_position(self, coords):
...
+    @property
+    def title_text(self):
+        """ Returns the title text of the card
+        """
+
+        return self.title_box.message
+
+    @title_text.setter
+    def title_text(self, text):
+        self.title_box.message = text
```
Comment (Nibba2018) [inline fury/ui.py:5404]: Rename `body_text` to `body` and `title_text` to `title` to make the API look more pythonic.

You should also implement the `resize` method

**Review by Nibba2018 — COMMENTED** [review]
Hello @antrikshmisri , Thank you for creating this PR. I have added some comments below, do have a look.

Thanks!

**Comment by antrikshmisri** [comment]
@Nibba2018 Sure, on it.

**Comment by antrikshmisri** [comment]
@Nibba2018 Also, for some reason any image with a `.png` extension is rendering in black and white. I assume it has something to do with fury's [`load_image`](https://github.com/fury-gl/fury/blob/58d45a4c64514924247226e6390b3c05b3703ce7/fury/io.py#L9) function or am I doing something wrong.

**Comment by antrikshmisri** [comment]
@Nibba2018 The [codecov/patch](https://github.com/fury-gl/fury/pull/398/checks?check_run_id=2213318229) is failing as the tests don't cover left click/drag function. Is it necessary to add these tests as these function are to be extracted to the base UI class?

**Comment by Nibba2018** [comment]
Hey @antrikshmisri , you need to implement those tests or else it will reduce the overall code coverage. You can look into other UI tests like `LineSlide2D`, `ComboBox2D` etc to understand how they work.

Just run those tests individually and you will get an idea.

**Comment by antrikshmisri** [comment]
> Hey @antrikshmisri , you need to implement those tests or else it will reduce the overall code coverage. You can look into other UI tests like `LineSlide2D`, `ComboBox2D` etc to understand how they work.
> 
> Just run those tests individually and you will get an idea.

Hey @Nibba2018 , will these tests suffice?

> Hey @antrikshmisri , you need to implement those tests or else it will reduce the overall code coverage. You can look into other UI tests like `LineSlide2D`, `ComboBox2D` etc to understand how they work.
> 
> Just run those tests individually and you will get an idea.

Hey @Nibba2018, will these tests suffice?

**Comment by Nibba2018** [comment]
Hey @antrikshmisri , I think you need to rebase this branch on top of master branch to trigger the codecov service.

**Comment by antrikshmisri** [comment]
@Nibba2018 I rebased the branch but I can't figure out why previously merged commits along with their diffs are being displayed. Should I create a new PR or is there a workaround for this?

**Comment by antrikshmisri** [comment]
@Nibba2018 I have rebased the branch but it still shows the same thing also for some reason the tests are failing, I looked into the logs and this import statement seems to be causing the failing checks.
```python
from .vtkRenderingOpenGL2 import *
```

File: docs/tutorials/02_ui/viz_card.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+panel.center = (500, 430)
+
+###############################################################################
+# Then we'll create two cards and add them to the panel.
+#
+# Note that here we specify the positions with floats. In this case, these are
+# percentages of the panel size.
+
+first_img_path = read_viz_icons(fname="circle-left.png")
+second_img_path = read_viz_icons(fname="circle-right.png")
```
Comment (skoudoro) [inline docs/tutorials/02_ui/viz_card.py:34]: Please, use different images. this is quite ugly 😅  and the resolution is bad

File: docs/tutorials/02_ui/viz_card.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+# First we need to fetch some icons that are included in FURY.
+
+fetch_viz_icons()
+
+###############################################################################
+# Let's create some cards and put them in a panel.
+# First we'll make the panel.
+
+panel = ui.Panel2D(size=(900, 800), color=(1.0, 1.0, 1.0))
+panel.center = (500, 430)
```
Comment (skoudoro) [inline docs/tutorials/02_ui/viz_card.py:25]: panel is not needed in this tutorial. You can keep it simpler

**Review by skoudoro — COMMENTED** [review]
Hi @antrikshmisri,

Thank you for this. See below my first comment. 

Please, check the Numpy doc style to update the overall docstring, there are many mistakes. Could you also improve the visual design of this UI (like adding a border+effect)?

**Comment by skoudoro** [comment]
Could you also rebase to fix the CI's. Thank you

**Comment by antrikshmisri** [comment]
@skoudoro Thanks for the review, I have fixed all the docstring w.r.t numpy docs. Just a few things that I wanted to know:-

1. The size of body need not to be larger than the title of the card. The sizes will be different only when the text in the body exceeds the allotted width, gets wrapped and is sent to the next line. Since FURY's `TextBlock2D` doesn't support this, I figured a solution to this would be to keep the sizes of title and body in a specific ratio (0.34 seemed to be a good ratio). Do you think this would suffice?
2. Is there a way to use an external image (i.e. not present in FURY's icon dataset) in the `ImageContainer2D`. This would help in making the card look better in the tutorial, as I agree it looks really ugly right now. One possibility would be to add support for fetching images from a URL in `ImageContainer2D`.
3. Regarding the border, I think this should be implemented under the `Panel2D` UI, that way it can be used in future UI's as well. What do you think ?

**Comment by skoudoro** [comment]
> The size of body need not to be larger than the title of the card. The sizes will be different only when the text in the body exceeds the allotted width, gets wrapped and is sent to the next line. Since FURY's TextBlock2D doesn't support this, I figured a solution to this would be to keep the sizes of title and body in a specific ratio (0.34 seemed to be a good ratio). Do you think this would suffice?

ok, I need to experiment and play with it, I will let you know. Could you also look @Nibba2018?

> Is there a way to use an external image (i.e. not present in FURY's icon dataset) in the ImageContainer2D. This would help in making the card look better in the tutorial, as I agree it looks really ugly right now. One possibility would be to add support for fetching images from a URL in ImageContainer2D.

I agree that we need to add this support. It would be better to do it in a new PR. Feel Free to start this PR and then, we can rebase this one.

> Regarding the border, I think this should be implemented under the Panel2D UI, that way it can be used in future UI's as well. What do you think ?

That's a good point and the same comment as above, we need a new PR for that. We might need to put this PR on standby until we fix these 2 issues.

**Comment by Nibba2018** [comment]
> The size of body need not to be larger than the title of the card. The sizes will be different only when the text in the body exceeds the allotted width, gets wrapped and is sent to the next line. Since FURY's `TextBlock2D` doesn't support this, I figured a solution to this would be to keep the sizes of title and body in a specific ratio (0.34 seemed to be a good ratio). Do you think this would suffice?

If I understand correctly, you want the Text to break lines once it overflows the width right?

Well in that case already have a method implemented called `clip_overflow` in `fury/ui.py` we reuse that code to implement that feature. Maybe we can add additional arguments to the TextBlock constructor to provude this feature.

**Comment by antrikshmisri** [comment]
> Well in that case we already have a method implemented called `clip_overflow` in `fury/ui.py` we can reuse that code to implement that feature. Maybe we can add additional arguments to the TextBlock constructor to provide this feature.
I guess this should solve the issue. I'll try to integrate it with the TextBlock UI.

**Comment by Nibba2018** [comment]
@antrikshmisri , Yes. I would suggest you to separate the part which checks for the overflow in a separate function like `is_overflowing` so that it can be used for other purposes. Let me know if you have any issues.

**Comment by antrikshmisri** [comment]
> Well in that case we already have a method implemented called `clip_overflow` in `fury/ui.py` we can reuse that code to implement that feature. Maybe we can add additional arguments to the TextBlock constructor to provide this feature.

@Nibba2018 I guess this function clips the text on width overflow. We need to wrap the text on overflow.

**Comment by Nibba2018** [comment]
> > Well in that case we already have a method implemented called `clip_overflow` in `fury/ui.py` we can reuse that code to implement that feature. Maybe we can add additional arguments to the TextBlock constructor to provide this feature.
> 
> @Nibba2018 I guess this function clips the text on width overflow. We need to wrap the text on overflow.

Yes, use the logic in this method to wrap the text on overflow. That's why I was hinting at creating a new method named `is_overflowing`.

**Comment by antrikshmisri** [comment]
> Yes, use the logic in this method to wrap the text on overflow. That's why I was hinting at creating a new method named `is_overflowing`.

Ah okay, got it.

**Comment by antrikshmisri** [comment]
Hey @skoudoro @Nibba2018, this PR has been rebased w.r.t the UI restructuring. Thanks!

**Comment by skoudoro** [comment]
Stalled:  waiting for #442 #441 to be merged

**Comment by antrikshmisri** [comment]
Hey @skoudoro @Nibba2018, support for border as well as clip/wrap overflow has been added, PTAL. Thanks!

**Comment by antrikshmisri** [comment]
Hey @skoudoro, this has been rebased and is ready for the next review. Thanks!

**Comment by ganimtron-10** [comment]
Hello @antrikshmisri ,
Can you please rebase this PR so we could move forward with this?

**Comment by antrikshmisri** [comment]
Hey @skoudoro, wanted to know why the `fetch_viz_icons` import was removed under this commit, couldn't find an explanation in the commit: https://github.com/skoudoro/fury/blame/d2bba5c511baf44dea9ae4e8f90e6b4a2b8c7878/fury/ui/tests/test_elements.py#L14? Is not necessary now to fetch the icon-set before using them?

**Comment by antrikshmisri** [comment]
Hey @ganimtron-10, resolved all the conflicts. The tests are failing but it is not related to my changes.

**Comment by skoudoro** [comment]
Hi @antrikshmisri,

> wanted to know why the fetch_viz_icons import was removed under this commit, couldn't find an explanation in the commit

I do not remember. I think it was not necessary because it was already called before. I need to check deeper 

Thank you for the rebase. @ganimtron-10 and me will do a last review to make sure everything is ok and we should merge this PR this weekend or next week.

**Comment by skoudoro** [comment]
Also, Can you do a small commit to add your tutorial in the file [`valid_examples.toml`](https://github.com/fury-gl/fury/blob/master/docs/examples/_valid_examples.toml#L68) @antrikshmisri ?

File: docs/examples/viz_card_sprite_sheet.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+###############################################################################
+# Converting numpy array sprites to vtk images
+sprite_to_vtk()
+
+###############################################################################
+# Adding a timer to update the card image
+show_manager.add_timer_callback(True, int(FRAME_TIME), timer_callback)
+
+# To interact with the UI, set interactive = True
+interactive = True
```
Comment (skoudoro) [inline docs/examples/viz_card_sprite_sheet.py:93]: it should be False

File: fury/ui/elements.py [hunk 14]
```diff
@@ -4103,3 +4105,265 @@ def _set_position(self, _coords):
...
+    ----------
+    image: :class: 'ImageContainer2D'
+        Renders the image on the card.
+    title_box: :class: 'TextBlock2D'
+        Displays the title on card.
+    body_box: :class: 'TextBLock2D'
+        Displays the body text.
+    """
+
+    def __init__(self, image_path, body_text="Body", draggable=True,
```
Comment (skoudoro) [inline fury/ui/elements.py:4123]: `body_text=""` or `body_text=None` instead

File: fury/ui/elements.py [hunk 15]
```diff
@@ -4103,3 +4105,265 @@ def _set_position(self, _coords):
...
+                 title_text="Title", padding=10, position=(0, 0),
```
Comment (skoudoro) [inline fury/ui/elements.py:4124]: `title_text=""` or `title_text=None`  instead

**Review by skoudoro — COMMENTED** [review]
Hi @antrikshmisri,

Overall, it looks good.  See below few comments then it is ready to be merged.

Do not forget to add your tutorial in `_valid_examples.toml`

Thanks in advance

File: docs/examples/viz_card.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+"""
+===============
+Card
+===============
```
Comment (ganimtron-10) [inline docs/examples/viz_card.py:5]: same as above

File: docs/examples/viz_card_sprite_sheet.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+"""
+===============
+Card
+===============
```
Comment (ganimtron-10) [inline docs/examples/viz_card_sprite_sheet.py:5]: same as above

**Review by ganimtron-10 — COMMENTED** [review]
Hey @antrikshmisri ,
@skoudoro has covered all the things.
I just had this below comment.

File: docs/examples/viz_card.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+"""
+===============
```
[the remark this answers is not in the document]
Comment (antrikshmisri) [reply docs/examples/viz_card.py]: I don't think the length of under/overline needs to be the same as the text but just for consistency's sake making the change.

File: docs/examples/viz_card.py (hunk 16 above)
Comment (antrikshmisri) [reply docs/examples/viz_card.py]: Replied above

File: docs/examples/viz_card_sprite_sheet.py (hunk 17 above)
Comment (antrikshmisri) [reply docs/examples/viz_card_sprite_sheet.py]: Replied above.

File: docs/examples/viz_card_sprite_sheet.py (hunk 13 above)
Comment (antrikshmisri) [reply docs/examples/viz_card_sprite_sheet.py]: Done.

File: fury/ui/elements.py (hunk 15 above)
Comment (antrikshmisri) [reply fury/ui/elements.py]: Done.

File: fury/ui/elements.py (hunk 14 above)
Comment (antrikshmisri) [reply fury/ui/elements.py]: Done.

**Comment by antrikshmisri** [comment]
@skoudoro @ganimtron-10 resolved all review comments. Added the examples to the `_valid_examples.toml` file

File: docs/examples/viz_card.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ganimtron-10) [reply docs/examples/viz_card.py]: Thanks for doing that!

**Comment by skoudoro** [comment]
All good, Thank you for the update @antrikshmisri and thank you for the review @ganimtron-10 

merging

**Pull request merged**